include/deal.II/lac/diagonal_matrix.h

```diff
@@ -403,6 +415,28 @@ DiagonalMatrix<VectorType>::Tvmult_add(VectorType &      dst,
 }
 
 
+
+template <typename VectorType>
+void
+DiagonalMatrix<VectorType>::apply_to_subrange(
+  const unsigned int index_of_first_unknown,
+  const unsigned int length,
+  const value_type * src,
+  value_type *       dst) const
+{
+  AssertIndexRange(index_of_first_unknown,
+                   diagonal.locally_owned_elements().n_elements());
+  AssertIndexRange(index_of_first_unknown + length,
+                   diagonal.locally_owned_elements().n_elements() + 1);
+
+  const value_type *diagonal_entry = diagonal.begin() + index_of_first_unknown;
+
+  DEAL_II_OPENMP_SIMD_PRAGMA
+  for (unsigned int i = 0; i < length; ++i)
+    dst[i] = diagonal_entry[i] * src[i];
```

Comment:
I am wondering if we should make some memory alignment assumptions here? Also shouldn't we add a pragma to indicate that `dst` and `src` are definitely not pointing to the same chunk in memory.

Reply:
I think the `DEAL_II_OPENMP_SIMD_PRAGMA` indicates that the regions do not overlap, so the compiler nicely vectorizes things. We do fulfill alignment requirements, too, but those are not needed for performance: The compiler will add unaligned load operations and in case the arrays are aligned, we will get full performance (in the x86 world this works on all newer processors since around 2014 I think). We can revise this in case we need it later.